### PR TITLE
Major update

### PR DESCRIPTION
**Pull request by STUkh**
- Updated Emscripten support to latest Emcc 1.39.11
- Updated LZ4 sources to 1.9.3
- Build WASM module by default
- Build ASM.js module
- Added support for more LZ4 options
- Updated build system to gulp v4
- Split all codebase to modules and rewrite them in `class` style
- Codebase build with Rollup
- Fixed syntax usage for cli, example, tests, etc.

### CODE REVIEW

File: example/bench.html [hunk 1]
```diff
@@ -1,31 +1,37 @@
...
+    <title>lz4.js bench</title>
+    <script src="../dev/lz4.js"></script>
+  </head>
+  <body>
+    <div>
+      <input type="file" id="f" />
+      <div id="result"></div>
+    </div>
+    <script>
+      const lz4init = lz4();
```
Comment (ukyo) [inline example/bench.html:13]: `lz4init` is defined by emacripten.

```
Uncaught SyntaxError: Identifier 'lz4init' has already been declared
```

File: bin/lz4-asm [hunk 2]
```diff
@@ -40,26 +60,28 @@ var argv = require('yargs')
 .epilog('If you want to show more info, refer to https://github.com/Cyan4973/lz4/blob/master/lz4_Frame_format.md')
 .argv
```
Comment (ukyo) [inline bin/lz4-asm:61]: Now, `argv` is a function call. Add `;`.

File: src/post.js [hunk 3]
```diff
@@ -1,296 +1,55 @@
...
-      BaseDecompressor.call(this);
-      Transform.call(this, {});
-      this.srcSize = 0;
-      this.dstSize = 0;
-      this.src = new Buffer(0);
-      this.dst = new Buffer(0);
+// Replace Thenable interface with Promise for Asynchronous modules
+if (Module['then']) {
+  delete Module['then'];
+  Module['ready'] = new Promise((resolve, reject) => {
```
Comment (ukyo) [inline src/post.js:41]: `Module['ready']` is used inside latest emscripten so it doesn't seem that `ready` is working properly. Simply call `lz4init()` to get the initialized module.

Here is a part of the code generated by emscripten.

```js
// Set up the promise that indicates the Module is initialized
var readyPromiseResolve, readyPromiseReject;
Module['ready'] = new Promise(function(resolve, reject) {
  readyPromiseResolve = resolve;
  readyPromiseReject = reject;
});
    
// ...

  return lz4init.ready
}
```

File: gulp/emscripten.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+function buildLib(cb) {
+  exec(`emmake make lib`, { cwd: 'lz4' }, cb);
+}
+
+function compileDev(cb) {
+  exec(`emcc -v ${INCLUDES} ${C_FILES} ${LIBS} -o dev/lz4.js ${POST_JS} ${DEV_ARGS.join(" ")} ${EXPORTED_FUNCTIONS}`, cb);
+}
+
+function compileAsmRelease(cb) {
+  exec(`emcc ${ASM_RELEASE_ARGS} ${INCLUDES} ${C_FILES} ${LIBS} -o tmp/_lz4.js ${POST_JS} ${EXPORTED_FUNCTIONS}`, cb);
```
[the remark this answers is not in the document]
Comment (STUkh) [reply gulp/emscripten.js:66]: Added

File: .gitignore [hunk 5]
```diff
@@ -1,2 +1,12 @@
...
+.DS_*
+*.sublime*
+.idea
+
+# Project
 node_modules/
 dev/
+test/_dst1.txt
+test/_dst2.txt
+lz4/
```
[the remark this answers is not in the document]
Comment (STUkh) [reply .gitignore]: Added

File: example/bench.html (hunk 1 above)
Comment (STUkh) [reply example/bench.html]: Fixed. Now looks good

File: bin/lz4-asm (hunk 2 above)
Comment (STUkh) [reply bin/lz4-asm]: Fixed

File: src/post.js (hunk 3 above)
Comment (STUkh) [reply src/post.js]: Refactored due to new module loading way.

**Comment by STUkh** [comment]
@ukyo added code due to requested changes 🚀

**Comment by ukyo** [comment]
@STUkh Thank you! LGTM 👍

**Pull request merged**